### ISSUE TITLE
Pre Aggregation is not working when timeDimensionReference is array

### ISSUE THREAD

**Issue by MuhammadMattiullah** [cube-js/cube#233]
I have added a pre-aggregation block into my schema file and mocked a specific query for pre-aggregation but didnt find any table that persist the table of aggregated tables. As mentioned in the documentation there are two ways for aggregation i.e in memory cache or db to store the aggregated data.
I am not getting where to allow the table permission as given below:
Cube.js should have write access to stb_pre_aggregations.(Documentation of Cube Js).
There's no clear stuff provided to configure Pre-Aggregations in cube js.
Kindly guide me how can i configure, access and use pre-aggregation in my cube js app.
Thanks in Advance

**Comment by paveltiunov** [comment]
@MuhammadMattiullah Hey Muhammad! Could you please share your schema file and query you expect to be pre-aggregated?

**Comment by MuhammadMattiullah** [comment]
Hello Pavel! Thanks for your response. 
Provided the schema file

cube(`SingleLevelEvents`, {
  sql: `SELECT * FROM test.\`SingleLevelEvents\``,
  
  joins: {
  
  },
  
  measures: {
 
  SumOfEnergy: {
  sql: `${CUBE}.\`data.energy\``,
  type: `sum`
  },
  AverageOfVoltage: {
  sql: `${CUBE}.\`data.voltage\``,
  type: `avg`
  },
  SumOfCurrent: {
  sql: `${CUBE}.\`data.current\``,
  type: `sum`
  },
  AverageOfAmbientTemperature: {
  sql: `${CUBE}.\`data.ambientTemperature\``,
  type: `avg`
  },
  SumOfPower: {
  sql: `${CUBE}.\`data.power\``,
  type: `sum`
  },
  
  },
  
  dimensions: {
  assetid: {
  sql: `${CUBE}.\`assetId\``,
  type: `string`
  },
  
  deviceid: {
  sql: `${CUBE}.\`deviceId\``,
  type: `string`
  },
  
  type: {
  sql: `type`,
  type: `string`
  },
  
  createdat: {
  sql: `${CUBE}.\`createdAt\``,
  type: `time`
  },
  
  updatedat: {
  sql: `${CUBE}.\`updatedAt\``,
  type: `time`
  },
  
  time: {
  sql: `time`,
  type: `time`
  }
  },
  preAggregations: {
  EnergyWrtAssetId: {
  type: `rollup`,
  measureReferences: [SingleLevelEvents.MaxEnergy],
  dimensionReferences: [SingleLevelEvents.assetid],
  timeDimensionReference: [SingleLevelEvents.createdat],
  granularity: `day`
  }
  }
  
  });

How can we verify that pre aggregation is working or not ? Bcz there's no mechanism described in doc to verify this because there's no aggregation tables in db.
Your quick response would be highly appreciated

**Comment by paveltiunov** [comment]
@MuhammadMattiullah Gotcha. Yeah. Time dimension should be referenced as
```
timeDimensionReference: SingleLevelEvents.createdat
```
instead of
```
timeDimensionReference: [SingleLevelEvents.createdat]
```
We can make this issue to track this validation bug.

**Comment by MuhammadMattiullah** [comment]
Hello @paveltiunov Thanks for your quick response.
I have tried the above solution as well. 
Can you please tell me that where i could find the pre-aggregated tables or how can i verify if pre-aggregation is working or not ?

**Comment by MuhammadMattiullah** [comment]
Updated Pre-aggregation Block in schema File:
  preAggregations: {
    EnergyWrtAssetId: {
      type: `rollup`,
      measureReferences: [SumOfEnergy],
      dimensionReferences: [assetid],
      timeDimensionReference: createdat,
      granularity: `day`
    }
   }
I got the below error when i try to run the query:

**Error: parse sql 'CREATE SCHEMA IF NOT EXISTS stb_pre_aggregations' error: unexpected SCHEMA at position 15 near schema**

But when i removed the pre-aggregation from the schema file then its working fine and results are expected but after adding the pre-aggregation block to my schema file then i got the error provided above..
Kindly guide me how can i achieve the pre-aggregation @paveltiunov

**Comment by antoniablair** [comment]
@MuhammadMattiullah Not sure if this is the cause of the error but the `createdat` in your code snippet is missing the column `SingleLevelEvents.createdat`

Normally to see if it is working correctly, you could try calling the CubeJS API with a query that should be pre-aggregated. This will cause it to create a pre-aggregation table if it does not exist and add a row, so you could see it by viewing all tables.

**Comment by paveltiunov** [comment]
@MuhammadMattiullah Which DB do you use?

**Comment by MuhammadMattiullah** [comment]
I am using mongoDb by using MongoDB ConnectorBI for queries. mongosqld for connecting with mongodb @paveltiunov

**Comment by paveltiunov** [comment]
@MuhammadMattiullah That totally makes sense. Unfortunately pre-aggregations currently aren't supported for MongoDB because MongoDB BI connector doesn't support table creation right now. There's an opportunity to support external pre-aggregations for MongoDB though. We need to tweak pre-aggregation creation process to respect MongoDB inability to create tables. I think you can create separate issue for that.

**Comment by MuhammadMattiullah** [comment]
@paveltiunov Thankyou so much for your quick response!
is there any workaround for this usecase ?
Will it work if i manually create tables ?

**Comment by paveltiunov** [comment]
@MuhammadMattiullah Unfortunately only viable option is external pre-aggregations. Cube.js currently doesn't support manually created tables for pre-aggregations.

**Comment by MuhammadMattiullah** [comment]
@paveltiunov can you please elaborate external pre-aggregation ?

**Comment by paveltiunov** [comment]
@MuhammadMattiullah Yep. Sure! Here's a doc for it: https://cube.dev/docs/pre-aggregations#external-pre-aggregations

**Comment by MuhammadMattiullah** [comment]
@paveltiunov please guide me the right direction that if i setup another mysql db for pre-aggregation so the pre-aggregation results will be stored in the mysql db and raw queries will be run via mongosqld ..
If i want to use external pre-aggregation with mysql then where to configure it because i have seen the document and the variables are passing to the sever instance via .evn file (dotenv) so should i declare all the required external db variables in the .env file that are given below just like the other's ?
CUBEJS_EXT_DB_HOST
CUBEJS_EXT_DB_NAME
CUBEJS_EXT_DB_PORT
CUBEJS_EXT_DB_USER
CUBEJS_EXT_DB_PASS
If i declare the external db(mysql) credentials to the .env file. Am i going to the right way ?

**Comment by MuhammadMattiullah** [comment]
Hello @paveltiunov i have tried to setup the external pre-aggregation using mysql. I have setup a mysql server and created a db name (stb_pre_aggregations) and initialized a server with the same method as discussed in the doc
 CubejsServerCore.create({
      externalDbType: 'mysql',
      externalDriverFactory: () => new MySQLDriver({
        host: 'localhost',
        database: 'stb_pre_aggregations',
        port: '3306',
        user: 'root',
        password: '1234',
      })
    })
Refrence Doc: https://cube.dev/docs/@cubejs-backend-server-core#options-reference-driver-factory  But after setting up this when i query the pre-aggregation column then it gives me the same error as given below:
Error: Error: parse sql 'CREATE SCHEMA IF NOT EXISTS stb_pre_aggregations' error: unexpected SCHEMA at position 15 near schema
Now i am using two db's one is mongo which is used for raw query and mysql db for pre-aggregation tables. Now setting up mysql db for pre-aggregations still i am getting the same error.
Please guide me how to achieve pre-aggregation.

**Comment by paveltiunov** [comment]
@MuhammadMattiullah Yep. It's right behavior. It's why we need #239 .

**Comment by MuhammadMattiullah** [comment]
@paveltiunov I have applied the same steps using mysql but still getting the same error.
Error: Error: parse sql 'CREATE SCHEMA IF NOT EXISTS stb_pre_aggregations' error: unexpected SCHEMA at position 15 near schema

**Comment by MuhammadMattiullah** [comment]
@paveltiunov please guide me that what else should be done in order to achieve pre-aggregation ? I will be waiting for your response.
Your quick responses are much appreciated!

**Comment by paveltiunov** [comment]
@MuhammadMattiullah What steps exactly have you done? This error is clearly from MongoDB driver. Again before pre-aggregations would be possible for MongoDB we need to implement #239.

**Comment by MuhammadMattiullah** [comment]
@paveltiunov I have setup mysql server and instantiating a core server like below:
CubejsServerCore.create({
externalDbType: 'mysql',
externalDriverFactory: () => new MySQLDriver({
host: 'localhost',
database: 'stb_pre_aggregations',
port: '3306',
user: 'root',
password: '1234',
})
})
All the credentials passed are mysql db.

**Comment by paveltiunov** [comment]
@MuhammadMattiullah I see. What's in your .env file?

**Comment by MuhammadMattiullah** [comment]
CUBEJS_DB_HOST=localhost
CUBEJS_DB_NAME=test
CUBEJS_DB_PORT=3307
CUBEJS_DB_TYPE=mongobi
CUBEJS_API_SECRET=ff564cb2631501b48bc7acef3427a7dcacad6317aacf53899440b7308b1a2f7eacea5312f750900f77551eb16539325b330bb46b38dbbcabcd0ce598b0a50c
This is the env file @paveltiunov

**Comment by paveltiunov** [comment]
@MuhammadMattiullah Yep. Right. You connect to MongoBI and it's why you see this error. External rollups for MongoBI DB *aren't supported yet*. You created #239 to track this support implementation.

**Comment by MuhammadMattiullah** [comment]
@paveltiunov as per our discussion i have mentioned above that now i m using two db's now
One is mongo db which is used for raw query that is configured in .env file and the second db is mysql db which i am using for pre-aggregation.

**Comment by paveltiunov** [comment]
@MuhammadMattiullah Yep. Right. And in this case pre-aggregations tables will be created in MongoDB first and only after that downloaded to MySQL because it's how it works right now. It's what we need to fix.

**Comment by MuhammadMattiullah** [comment]
@paveltiunov  I am setting up mysql for the implementation of pre-aggregation only and mongo db is used for raw queries. I configured mongo db in .env file like below:

CUBEJS_DB_HOST=localhost
CUBEJS_DB_NAME=test
CUBEJS_DB_PORT=3307
CUBEJS_DB_TYPE=mongobi
CUBEJS_API_SECRET=ff564cb2631501b48bc7acef3427a7dcacad6317aacf53899440b7308b1a2f7eacea5312f750900f77551eb16539325b330bb46b38dbbcabcd0ce598b0a50c

i configured mysql db for pre-aggregation like below:
CubejsServerCore.create({
externalDbType: 'mysql',
externalDriverFactory: () => new MySQLDriver({
host: 'localhost',
database: 'stb_pre_aggregations',
port: '3306',
user: 'root',
password: '1234',
})
})

**Comment by MuhammadMattiullah** [comment]
@paveltiunov so that means i can't use pre-aggregation at this moment ?

**Comment by paveltiunov** [comment]
@MuhammadMattiullah Yes. Need to implement #239 first.

**Comment by MuhammadMattiullah** [comment]
@paveltiunov can you please tell me that how much time will it take to implement this feature ?

**Comment by igorlukanin** [comment]
Let me make a summary of this thread:

https://github.com/cube-js/cube/issues/239 was resolved https://github.com/cube-js/cube/pull/395. meaning that it is not possible to create pre-aggregations from MongoDB without writing the data to Mongo.

Also, Pavel has provided the solution for the initial issue in this comment: https://github.com/cube-js/cube/issues/233#issuecomment-545103162

It looks like this one is fully resolved then. Closing the issue, feel free to reopen if needed.
